### DISCUSSION TITLE
Event Fired on Context Update

### DISCUSSION THREAD

**Discussion by dhirjgupta** [statelyai/xstate#2253] [Ideas]
An Idea I've raised in Discord, but I thought it'd be good to have a dedicated space to debate / refine this and related ideas.

Idea: I think it would be useful if we could set a flag when instantiating a machine so that it sends itself an event every time context updates. This would be useful in a variety of situations, including the following: 

If I have two parameters stored in a machine context, and when either of them change, I want to query an API for fresh data. In that system, there are many different events that update those values.
Currently,  I am required to remember to issue an api request event wherever I write an assign to one of those values...
```js
on: {
  EVENT_1: {actions: [
    assign({param1: (c,e) => e.payload },
    send({ //repeated send
      type: "TRIGGER_API_QUERY_A"
      params: {
        param1: event.payload
        param2: context.param2
    })
  ]},
  EVENT_2: {actions: [
    assign({param2: (c,e) => e.payload },
    send({ //repeated send
      type: "TRIGGER_API_QUERY_A"
      params: {
        param1: context.param2
        param2: event.payload
    })
  ]},
  ///... any number of other events that update param1 or param2
}
```
To maintain my codebase, I have to keep track of every event that touches those params, and then repeat myself in every spot.
However with some kind of "context changed" event...

```js
on: {
  [CONTEXT_CHANGED_EVENT]: {
    actions: [
      pure((context, event) => {
         const updatedKeys: string[] = event.updatedKeys;
         if (updatedKeys.includes('param1') || updatedKeys.includes('param2')) {
            return [
              send({
                type: "TRIGGER_API_QUERY_A"
                params: {
                  param1: context.param1
                  param2: context.param2
                }
            })
         } else {
           return []
         }
      })
    ]
  }
}
```

This functionality would allow a user to better encode what my machine is really reacting to:  param1 or param2 changed. For this particular piece of logic, I don't care if it's because the event was due to RESET_TO_DEFAULT event,  USER_INPUT_1 event, USER_INPUT_5 event,  or URL_CHANGE event. Those aren't what we're really responding to.

**Comment by fraserdrops** [comment]
This is an interesting idea that I've thought about too. I came to the same conclusion as you: if you want to do something when context changes the way to do this currently is `actions: ['updateContext', 'doSomethingWithUpdatedContext']`.

Now the bit I'm not sure agree with is 'This functionality would allow a user to better encode what my machine is really reacting to'. 

There's two ways you could think of the chain of events:
1. When `EVENT_1` event happens, update context and send/trigger api query. How it currently works now.
2. When `EVENT_1` event happens, update context. When `CONTEXT_CHANGED_EVENT` event happens, check if param1 or param2 has changed, if one of them has then send/trigger api query. This is how you're thinking about it, if I have it right

I think there's a few clues that approach 1 is perhaps a nicer way to model it.

a) If you look at the context changed event, you'll see that you have to check if the updatedKeys includes param1 or param2:
```javascript
on: {
  [CONTEXT_CHANGED_EVENT]: {
    actions: [
      pure((context, event) => {
         const updatedKeys: string[] = event.updatedKeys;
         if (updatedKeys.includes('param1') || updatedKeys.includes('param2')) {
            return [
              send({
                type: "TRIGGER_API_QUERY_A"
                params: {
                  param1: context.param1
                  param2: context.param2
                }
            })
         } else {
           return []
         }
      })
    ]
  }
}
```
But back here:
```javascript
EVENT_2: {actions: [
    assign({param2: (c,e) => e.payload },
  ]}
```
we _knew_ that one of those params was being updated. By adding the step of waiting for `CONTEXT_CHANGED_EVENT` event, we've lost information about the situation that we had when the original `EVENT_1` event was handled. 

b) What if you sometimes want to update the params without triggering a query? Maybe you want to batch queries or something. With approach 1, you can just not trigger the action to query api when you're in the BATCHING state. With approach 2, you'd most likely need to handle the `CONTEXT_CHANGED_EVENT` in each state. This seems like a bit of effort, essentially handling two events instead of one. Is needing to make sure that the `CONTEXT_CHANGED_EVENT` is handled correctly in each state better than having to remember to fire a second action after the update?

I don't mean to sound dismissive of this idea, as I said I have considered it before. And I've often wanted a way for XState to handle dervied data from context, and in theory `CONTEXT_CHANGED_EVENT` could have been a way to do that. I'm curious what everyone else thinks about it.
